A75-xds-aggregate-cluster-behavior-fixes.md

```diff
@@ -0,0 +1,312 @@
+    protobuf, we NACK.
+  - Otherwise, we look at the
+    [`common_http_protocol_options`](https://github.com/envoyproxy/envoy/blob/9108a1457513cdad1881fc6177a008aa511ef242/api/envoy/extensions/upstreams/http/v3/http_protocol_options.proto#L133)
+    field:
+    - If the field is not present, we assume the default value for
+      `idle_timeout`.
+    - If the field is present, we look at the
+      [`idle_timeout`](https://github.com/envoyproxy/envoy/blob/9108a1457513cdad1881fc6177a008aa511ef242/api/envoy/config/core/v3/protocol.proto#L240)
+      field:
+      - If the field is unset, we assume the default value for
+        `idle_timeout`.
+      - Otherwise, as with all `google.protobuf.Duration` fields, we
+        check the following:
+        - If the `seconds` field is outside of the range [0, 315576000000],
+          we NACK.
+        - If the `nanos` field is outside of the range [0, 999999999],
+          we NACK.
+        - Otherwise, we use the specified value.
+
+The `idle_timeout` defaults to 1 hour.
+
+The XdsClient will return the `idle_timeout` value as a member of the
+parsed CDS resource.
+
+#### Changes in xds_override_host LB Policy
+
+We will make the following changes in the xds_override_host policy:
+
+- The xds_override_host policy will take ownership of any subchannel
+  that the child policy does not own if needed for session affinity.
+- Each entry in the subchannel map will contain a `last_used_time` field,
+  which will indicate the last time at which the entry was used for a session.
+  Note that this field will be updated only when the xds_override_host picker
+  sees the attribute from the stateful session HTTP filter; if that filter
+  is not present, then SSA is not being used for the request, and we do not
+  need to retain ownership of the subchannel.
+- The xds_override_host policy will have a timer that runs periodically
+  to do a sweep over the subchannel map and unref any owned subchannel
+  whose `last_used_time` is older than the `idle_timeout` value from the
+  CDS resource.  The timer can initially be set for `idle_timeout`
+  duration, and subsequent runs can dynamically determine when to run
+  based on the `last_used_time` of the entries in the map, with a
+  minimum of 5 seconds between runs to minimize overhead.
+- When the child policy unrefs a subchannel it owns, if the entry's
+  `last_used_time` is newer than the `idle_timeout`, the
+  xds_override_host policy assumes ownership of the subchannel instead
+  of deleting it.
+- In the picker, when we iterate through the addresses in the cookie, if
+  we do not find any entries with a subchannel is states READY, IDLE, or
+  CONNECTING and we do find entries that do not have any subchannel, we
+  will trigger creation of a subchannel for one of those entries.
+
+The picker logic will now look like this:
+
+```
```

Comment:
nit
```suggestion
```py
```

Reply:
This is not actually intended to be valid python code; it's just python-style pseudo-code.  I've add a parenthetical indicating this.

Reply:
Up to you - even if it's not python, it'll still be easier to ready with syntax highlighting. It's clear now from your clarification it's pseudo-code, IMO adding `py` just for readability might worth it.